engine/src/main/java/org/pentaho/di/trans/steps/webservices/WebService.java

```diff
@@ -0,0 +1,1122 @@
+                  .getString( PKG, "WebServices.Log.ResultRowDataFound", nodeXML.toString() ) );
+              }
+
+              // Allocate a new row...
+              //
+              outputRowData = createNewRow( rowData );
+
+              // Let's see what's in there...
+              //
+              NodeList childNodes = node.getChildNodes();
+              for ( int j = 0; j < childNodes.getLength(); j++ ) {
+                Node childNode = childNodes.item( j );
+
+                field = meta.getFieldOutFromWsName( childNode.getNodeName(), ignoreNamespacePrefix );
+                if ( field != null ) {
+
+                  if ( getNodeValue( outputRowData, childNode, field, transformer, false ) ) {
+                    // We found a match.
+                    // This means that we are dealing with a single row
+                    // It also means that we need to update the output index pointer
+                    //
+                    fieldsFound++;
+                  }
+                }
+              }
+
+              // Prevent empty rows from being sent out.
+              //
+              if ( fieldsFound > 0 ) {
+                // Send a row in a series of rows on its way.
+                //
+                putRow( data.outputRowMeta, outputRowData );
+              }
+            }
+          }
+        }
+      }
+
+      if ( singleRow && fieldsFound > 0 ) {
+        // Send the single row on its way.
+        //
+        putRow( data.outputRowMeta, outputRowData );
+      }
+    } catch ( Exception e ) {
+      throw new KettleStepException( BaseMessages.getString(
+        PKG, "WebServices.ERROR0010.OutputParsingError", response.toString() ), e );
+    }
+  }
+
+  private Object[] createNewRow( Object[] inputRowData ) {
+    return inputRowData == null ? RowDataUtil.allocateRowData( data.outputRowMeta.size() ) : RowDataUtil
+      .createResizedCopy( inputRowData, data.outputRowMeta.size() );
+  }
+
+  private void compatibleProcessRows( InputStream anXml, Object[] rowData, RowMetaInterface rowMeta,
+                                      boolean ignoreNamespacePrefix, String encoding ) throws KettleException {
+
+    // First we should get the complete string
+    // The problem is that the string can contain XML or any other format such as HTML saying the service is no longer
+    // available.
+    // We're talking about a WEB service here.
+    // As such, to keep the original parsing scheme, we first read the content.
+    // Then we create an input stream from the content again.
+    // It's elaborate, but that way we can report on the failure more correctly.
+    //
+    String response = readStringFromInputStream( anXml, encoding );
+
+    // Create a new reader to feed into the XML Input Factory below...
+    //
+    StringReader stringReader = new StringReader( response );
+
+    try {
+      XMLInputFactory vFactory = XMLParserFactoryProducer.createSecureXMLInputFactory();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Parsing XML input using a weakly configured XML parser can lead to XML external entity (XXE) injection. To protect against XXE attacks, be sure to securely configure your XML parser.

[Learn more](https://skb.highcastle.a2z.com/implementations/189)